src/java.base/share/classes/java/lang/runtime/Carrier.java

```diff
@@ -0,0 +1,1120 @@
+     */
+    private static class CarrierArrayFactory {
+        /**
+         * Unsafe access.
+         */
+        private static final Unsafe UNSAFE;
+
+        static {
+            try {
+                UNSAFE = Unsafe.getUnsafe();
+                Lookup lookup = MethodHandles.lookup();
+                CONSTRUCTOR = lookup.findConstructor(CarrierArray.class,
+                        methodType(void.class, int.class, int.class));
+                GET_LONG = lookup.findVirtual(CarrierArray.class, "getLong",
+                        methodType(long.class, int.class));
+                PUT_LONG = lookup.findVirtual(CarrierArray.class, "putLong",
+                        methodType(CarrierArray.class, int.class, long.class));
+                GET_INTEGER = lookup.findVirtual(CarrierArray.class, "getInteger",
+                        methodType(int.class, int.class));
+                PUT_INTEGER = lookup.findVirtual(CarrierArray.class, "putInteger",
+                        methodType(CarrierArray.class, int.class, int.class));
+                GET_OBJECT = lookup.findVirtual(CarrierArray.class, "getObject",
+                        methodType(Object.class, int.class));
+                PUT_OBJECT = lookup.findVirtual(CarrierArray.class, "putObject",
+                        methodType(CarrierArray.class, int.class, Object.class));
+            } catch (ReflectiveOperationException ex) {
+                throw new AssertionError("carrier static init fail", ex);
+            }
+        }
+
+        /*
+         * Constructor accessor MethodHandles.
+         */
+        private static final MethodHandle CONSTRUCTOR;
+        private static final MethodHandle GET_LONG;
+        private static final MethodHandle PUT_LONG;
+        private static final MethodHandle GET_INTEGER;
+        private static final MethodHandle PUT_INTEGER;
+        private static final MethodHandle GET_OBJECT;
+        private static final MethodHandle PUT_OBJECT;
+
+        /**
+         * Wrapper object for carrier arrays. Instances types are stored in the {@code objects}
+         * array, while primitive types are recast to {@code int/long} and stored in the
+         * {@code primitives} array. Primitive byte, short, char, boolean and int are stored as
+         * integers. Longs and doubles are stored as longs.  Longs take up the first part of the
+         * primitives array using normal indices. Integers follow using int[] indices offset beyond
+         * the longs using unsafe getInt/putInt.
+         */
+        private static class CarrierArray {
+            /**
+             * Carrier for primitive values.
+             */
+            private final long[] primitives;
+
+            /**
+             * Carrier for objects;
+             */
+            private final Object[] objects;
+
+            /**
+             * Constructor.
+             *
+             * @param primitiveCount  slot count required for primitives
```

Comment:
This isn’t the slot count, but the array size, which is half the slot count for `long`s.

Reply:
Thank for pointing this out. It is the slot count, but I was over-allocating the long array.

```
diff --git a/src/java.base/share/classes/java/lang/runtime/Carrier.java b/src/java.base/share/classes/java/lang/runtime/Carrier.java
index 50a32346417..775df6cb4c3 100644
--- a/src/java.base/share/classes/java/lang/runtime/Carrier.java
+++ b/src/java.base/share/classes/java/lang/runtime/Carrier.java
@@ -331,7 +331,7 @@ public final class Carrier {
              * @param objectCount     slot count required for objects
              */
             CarrierArray(int primitiveCount, int objectCount) {
-                this.primitives = new long[primitiveCount];
+                this.primitives = new long[primitiveCount + 1 / 2];
                 this.objects = new Object[objectCount];
             }
 
@@ -432,10 +432,10 @@ public final class Carrier {
             int longCount = carrierShape.longCount();
             int intCount = carrierShape.intCount();
             int objectCount = carrierShape.objectCount();
-            int primitiveSlots = longCount * LONG_SLOTS + intCount;
+            int primitiveCount = longCount * LONG_SLOTS + intCount;
 
             MethodHandle constructor = MethodHandles.insertArguments(CONSTRUCTOR,
-                    0, primitiveSlots, objectCount);
+                    0, primitiveCount, objectCount);
 
             // long array index
             int index = 0;
```